### ISSUE TITLE
live reload doesnt work at all

### ISSUE THREAD

**Issue by shekspir55** [angular-fullstack/generator-angular-fullstack#2063]
[ ] I understand that GitHub issues are not for tech support, but for questions specific to this generator, bug reports, and feature requests.

Item | Version
generator-angular-fullstack | 3.7.5
Node | 4.4.7 6
npm | 3.10.3
Operating System | Ubuntu 14.04

Item | Answer
Babel
HTML
LESS
Router | ui-router
Build Tool | Grunt and Gulp
Client Tests | Jasmine / Mocha
DB | MongoDB
Auth | Y 

Livereload doesn't work neither with grunt nor with gulp. Changing client isn't affecting anything in the browser, but changing server reloads page on 404(I think that the page is reloading is being faster then server reload). After reload changes are applied. Is there a way of troubleshooting this problem?

**Comment by Awk34** [comment]
I haven't heard of anyone else running into this

**Comment by shekspir55** [comment]
I'm running server via `gulp serve`, I have tried installing generator on mac, but LiveReload still doesn't work. Last versions of generator work perfectly. I've created [blank repo](https://github.com/shekspir55/test-angular-fullstack) and I wonder if there live reload will work for you. Also I've tried with angular-fullstack 4.0.0 and different versions of node. But can't find the actual problem.

**Comment by shekspir55** [comment]
@Awk34 Hello, I've added this code before closing `</boody>`
`<script>//<![CDATA[
document.write('<script src="//' + (location.hostname || 'localhost') + ':35729/livereload.js?snipver=1"><\/script>')
//]]></script>`
of course this isn't best solution but it worked, in every project, but I don't understand why this isn't adding automatically as in previous builds.
Either way, thank you for this great generator.

**Comment by Awk34** [comment]
The project uses [connect-livereload](https://github.com/intesso/connect-livereload) in the Express config [here](https://github.com/angular-fullstack/generator-angular-fullstack/blob/3.7.5/templates/app/server/config/express.js#L89). That middleware should be adding the livereload snippet to your HTML. Is your `NODE_ENV` getting set to something other than `"development"` maybe?

**Comment by shekspir55** [comment]
I've put `console.log('envirement :', process.env.NODE_ENV);`
and got `envirement : development`
hove you got chance to check my repo?
Also `'development' === env` returns true.
 I think there is some problem with injection in html.

**Comment by Awk34** [comment]
@shekspir55 I probably won't have time in the near future, but maybe if you bug me later this week/weekend I can check it out

**Comment by shekspir55** [comment]
@Awk34 ok, I will remind you here.

**Comment by Koslun** [comment]
If you are ok with using 4.0 you can also try using the #2073 PR which replaces livereload with browser-sync.

You would however have to clone the repo, switch to the PR's branch and use `npm link` to use the cloned project as your local version of the generator. Or just copy + paste the changes to a fresh 4.0 generated project.

**Comment by bogini** [comment]
@shekspir55 I was having the same issue. The livereload script wasn't been added to the end of `index.html`. For some reason, the `connect-livereload` middleware was never called. 

Fixed it by modifying `server/config/express.js` and placing this code:

```
  if ('development' === env) {
    app.use(require('connect-livereload')({
      ignore: [
        /^\/api\/(.*)/,
        /\.js(\?.*)?$/, /\.css(\?.*)?$/, /\.svg(\?.*)?$/, /\.ico(\?.*)?$/, /\.woff(\?.*)?$/,
        /\.png(\?.*)?$/, /\.jpg(\?.*)?$/, /\.jpeg(\?.*)?$/, /\.gif(\?.*)?$/, /\.pdf(\?.*)?$/
      ]
    }));
  }
```

Above the the line (36):

`app.use(express.static(app.get('appPath')));`

Restart `gulp serve` and it should work as expected.

**Comment by shekspir55** [comment]
@bogini I don't know how but it actually solved the problem, thank you very much. Regards.

**Comment by bogini** [comment]
I just realised that my solution above breaks some routes (try for example accessing [http://localhost:9000/login](http://localhost:9000/login) directly).

**Comment by shekspir55** [comment]
@bogini yes, there are some routing issues.

**Comment by MarinAngelo** [comment]
Same problem here, live reload doesn't work at all. Isn't there a solution that works without routing problems?

**Comment by Dan466** [comment]
You need to move the line   

`app.set('appPath', path.join(config.root, `'client'));`
`
above `    app.use(express.static(path.join(config.root, '.tmp')));
` otherwise live reload will not work correctly. 

See Below: 

`/**
 * Express configuration
 */

'use strict';

import express from 'express';
import favicon from 'serve-favicon';
import morgan from 'morgan';
import compression from 'compression';
import bodyParser from 'body-parser';
import methodOverride from 'method-override';
import cookieParser from 'cookie-parser';
import errorHandler from 'errorhandler';
import path from 'path';
import lusca from 'lusca';
import config from './environment';
import session from 'express-session';

export default function(app) {
  var env = app.get('env');

  app.set('appPath', path.join(config.root, 'client'));

  if (env === 'development' || env === 'test') {
    app.use(express.static(path.join(config.root, '.tmp')));
  }

  if (env === 'production') {
    app.use(favicon(path.join(config.root, 'client', 'favicon.ico')));
  }

  app.use(express.static(app.get('appPath')));
  app.use(morgan('dev'));

  app.set('views', config.root + '/server/views');
  app.engine('html', require('ejs').renderFile);
  app.set('view engine', 'html');
  app.use(compression());
  app.use(bodyParser.urlencoded({ extended: false }));
  app.use(bodyParser.json());
  app.use(methodOverride());
  app.use(cookieParser());

  // Persist sessions with MongoStore / sequelizeStore
  // We need to enable sessions for passport-twitter because it's an
  // oauth 1.0 strategy, and Lusca depends on sessions
  app.use(session({
    secret: config.secrets.session,
    saveUninitialized: true,
    resave: false
  }));

  /**
   * Lusca - express server security
   * https://github.com/krakenjs/lusca
   */
  if (env !== 'test' && !process.env.SAUCE_USERNAME) {
    app.use(lusca({
      csrf: {
        angular: true
      },
      xframe: 'SAMEORIGIN',
      hsts: {
        maxAge: 31536000, //1 year, in seconds
        includeSubDomains: true,
        preload: true
      },
      xssProtection: true
    }));
  }

  if ('development' === env) {
    app.use(require('connect-livereload')({
      ignore: [
        /^\/api\/(.*)/,
        /\.js(\?.*)?$/, /\.css(\?.*)?$/, /\.svg(\?.*)?$/, /\.ico(\?.*)?$/, /\.woff(\?.*)?$/,
        /\.png(\?.*)?$/, /\.jpg(\?.*)?$/, /\.jpeg(\?.*)?$/, /\.gif(\?.*)?$/, /\.pdf(\?.*)?$/
      ]
    }));
  }

  if ('development' === env || 'test' === env) {
    app.use(errorHandler()); // Error handler - has to be last
  }
}
`

**Comment by nguyenct7** [comment]
I've noticed that Live Reload works when you first go to a route (i.e. http://localhost/about). But once you get back to the root, it stops working. Haven't come up with solution. I do plan on replacing it with BrowserSync, though.

**Comment by Khachatour** [comment]
Yes mate, @shekspir55 said that this line of code at the end of the body tag of index.html will solve that, specifically it solved for me 
```<script>
       document.write('<script src="//' + (location.hostname || 'localhost') + ':35729/livereload.js?snipver=1">       <\/script>')
</script>```
